src/qutip_tensorflow/core/data/tftensor.py

```diff
@@ -0,0 +1,98 @@
+import tensorflow as tf
+from tensorflow.errors import InvalidArgumentError
+import qutip
+import numbers
+
+__all__ = ["TfTensor"]
+
+
+class TfTensor(qutip.core.data.Data):
+    """This class provide a wraps around TensorFlow's Tensor. It will store data
+    as a Tensor of dtype `tensorflow.complex128`. Data will be expanded into a 2D
+    Tensor.
+
+    Parameters
+    ----------
+    data: array-like
+        Data to be stored. Accepts same array-like as `tensorflow.constant`.
+    shape: (int, int)
+        Shape of data. Default `None`, it tries to infer the shape from data
+        accessing the attribute `data.shape`.
+    copy: bool
+        If `True` (default) then the object is copied. Otherwise, copy will only
+        be made if tf.constant returns a copy of data(when input is not a
+        `Tenor`) or if a copy is needed to satisfy dtype (tensorflow.complex128) and
+        shape (2D Tensor)."""
+
+    def __init__(self, data, shape=None, copy=False):
+        # If the input is a tensor this does not copy it. Otherwise it will
+        # return a copy.
+        data = tf.constant(data)
+
+        # If dtype of Tensor is already a tf.complex128 then this will not
+        # return a copy
+        data = tf.cast(data, tf.complex128)
+
+        # Try to inherit shape from data and expand shape
+        if shape is None:
+            try:
+                shape = data.shape
+            except AttributeError:
+                raise ValueError(
+                    """Shape could not be inferred from data.
+                                 Please, include the shape of data."""
+                )
```

Comment:
`data` is now guaranteed to be a tensor, so is it ever possible for this to raise an exception?

Reply:
You are right, this was necessary for the previous logic but is not any more. Removing.